Lib/test/test_frame.py

```diff
@@ -431,6 +432,116 @@ def test_is_mapping(self):
                 kind = "other"
         self.assertEqual(kind, "mapping")
 
+    def test_proxy_key_stringlikes(self):
+        class StringSubclass(str):
+            pass
+
+        class ImpostorX:
+            def __hash__(self):
+                return hash('x')
+
+            def __eq__(self, other):
+                return other == 'x'
+
+        def f(obj):
+            x = 1
+            proxy = sys._getframe().f_locals
+            proxy[obj] = 2
+            return (
+                proxy.keys() ==  ['obj', 'x', 'proxy'],
+                proxy == {'obj': 'x', 'x': 2, 'proxy': proxy},
+                proxy
+            )
+
+        for obj in StringSubclass('x'), ImpostorX():
+            with self.subTest(cls=type(obj).__name__):
+
+                assertion1, assertion2, proxy = f(obj)
+                self.assertEqual(proxy.keys(),  ['obj', 'x', 'proxy'])
+                self.assertEqual(proxy, {'obj': 'x', 'x': 2, 'proxy': proxy})
+                self.assertTrue(assertion1)
+                self.assertTrue(assertion2)
+
+    def test_proxy_key_unhashables(self):
+        class StringSubclass(str):
+            __hash__ = None
+
+        class ObjectSubclass:
+            __hash__ = None
+
+        proxy = sys._getframe().f_locals
+
+        for obj in StringSubclass('x'), ObjectSubclass():
+            with self.subTest(cls=type(obj).__name__):
+                with self.assertRaises(TypeError):
+                    proxy[obj]
+                with self.assertRaises(TypeError):
+                    proxy[obj] = 0
+
+
+class FrameLocalsProxyMappingTests(mapping_tests.TestHashMappingProtocol):
+    """Test that FrameLocalsProxy behaves like a Mapping (with exceptions)"""
+
+    def _f(*args, **kwargs):
+        def _f():
+            return sys._getframe().f_locals
+        return _f()
+    type2test = _f
+
+    @unittest.skipIf(True, 'Unlike a mapping: empty proxy != empty proxy')
```

Comment:
```suggestion
    @unittest.skipIf(True, 'Locals proxies for different frames never compare as equal')
```

The skip message here is a little misleading.

Two empty proxies for the same frame do compare as equal (that's in the PEP, with even the currently published version without any of the beta clarifications containing the sentence: "`f.f_locals == f.f_locals` will be True")

The reason this mapping test fails is that for this to happen, the locals proxy instances must reference the *same* frame:

```
>>> def f():
...     return sys._getframe().f_locals, sys._getframe().f_locals
...
>>> empty1, empty2 = f()
>>> empty1 is empty2
False
>>> empty1 == empty2
True
>>> (empty1, empty2) == f()
False
```